libraries/chain_kv/include/b1/session/cache.hpp

```diff
@@ -0,0 +1,454 @@
+    auto contains(const bytes& key) const -> bool;
+    auto erase(const bytes& key) -> void;
+
+    template <typename iterable>
+    auto read(const iterable& keys) const -> const std::pair<std::vector<key_value>, std::unordered_set<bytes>>;
+
+    template <typename iterable>
+    auto write(const iterable& key_values) -> void;
+
+    template <typename iterable>
+    auto erase(const iterable& keys) -> void;
+
+    template <typename data_store, typename iterable>
+    auto write_to(data_store& ds, const iterable& keys) const -> void;
+
+    template <typename data_store, typename iterable>
+    auto read_from(const data_store& ds, const iterable& keys) -> void;
+    
+    auto find(const bytes& key) -> iterator;
+    auto find(const bytes& key) const -> const_iterator;
+    auto begin() -> iterator;
+    auto begin() const -> const_iterator;
+    auto end() -> iterator;
+    auto end() const -> const_iterator;
+    auto cbegin() const -> const_iterator;
+    auto cend() const -> const_iterator;
+    auto lower(const bytes& key) -> iterator;
+    auto lower(const bytes& key) const -> const_iterator;
+    auto upper(const bytes& key) -> iterator;
+    auto upper(const bytes& key) const -> const_iterator;
+    auto clower(const bytes& key) const -> const_iterator;
+    auto cupper(const bytes& key) const -> const_iterator;
+    
+    auto memory_allocator() -> const std::shared_ptr<allocator>&;
+
+private:
+    template <typename on_cache_hit, typename on_cache_miss>
+    auto find_(const bytes& key, const on_cache_hit& cache_hit, const on_cache_miss& cache_miss) const -> const key_value&;
+    
+private:
+    cache_type m_cache;
+    std::shared_ptr<allocator> m_allocator;
+};
+
+template <typename allocator>
+auto make_cache(std::shared_ptr<allocator> a) -> cache<allocator>
+{
+    return {std::move(a)};
+}
+
+template <typename allocator>
+cache<allocator>::cache(std::shared_ptr<allocator> memory_allocator)
+: m_allocator{std::move(memory_allocator)}
+{
+}
+
+// Searches the cache for the given key.
+//
+// \tparam on_cache_hit A functor to invoke when the key is found.  It has the following signature void(cache_type::iterator& it)
+// \tparam on_cache_miss A functor to invoke when the key is not found.  It has the following singature void(void).
+// \param key The key to search the cache for.
+// \param cache_hit The functor to invoke when the key is found.
+// \param cache_miss The functor to invoke when the key is not found.
+// \return A reference to the key_value instance if found, key_value::invalid otherwise.
+template <typename allocator>
+template <typename on_cache_hit, typename on_cache_miss>
+auto cache<allocator>::find_(const bytes& key, const on_cache_hit& cache_hit, const on_cache_miss& cache_miss) const -> const key_value&
+{
+    auto it = m_cache.find(key);
+    
+    if (it == std::end(m_cache))
+    {
+        cache_miss();
+        return key_value::invalid;
+    }
+    
+    if (cache_hit(it))
```

Comment:
When does cache_hit() return false?

Reply:
I saw it returns false when it is erased inside cache_hit().  it may need a comment.